File: subprojects/lang-ide-common/src/main/java/hu/bme/mit/semantifyr/lang/ide/server/wire/VerificationMetrics.java
/*
 * SPDX-FileCopyrightText: 2026 The Semantifyr Authors
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package hu.bme.mit.semantifyr.lang.ide.server.wire;

import hu.bme.mit.semantifyr.verifier.VerificationMetricsDto;
import java.time.Duration;

public record VerificationMetrics(
        Duration totalDuration,
        Duration preparationDuration,
        Duration verificationDuration,
        Duration backAnnotationDuration) {

    public static VerificationMetrics fromDto(VerificationMetricsDto dto) {
        var backend = dto.getBackend();
        var verifier = dto.getVerifier();

        var preparation = verifier.getCompilationDuration();
        if (backend != null) {
            preparation = preparation.plus(backend.getPreparationDuration());
        }

        var verification = backend != null ? backend.getVerificationDuration() : Duration.ZERO;

        var backAnnotation = verifier.getBackAnnotationDuration();
        if (backend != null) {
            backAnnotation = backAnnotation.plus(backend.getBackAnnotationDuration());
        }

        return new VerificationMetrics(dto.getTotalDuration(), preparation, verification, backAnnotation);
    }
}

File: subprojects/verifier/src/main/kotlin/hu/bme/mit/semantifyr/verifier/VerificationMetrics.kt
/*
 * SPDX-FileCopyrightText: 2026 The Semantifyr Authors
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package hu.bme.mit.semantifyr.verifier

import hu.bme.mit.semantifyr.backend.BackendMetrics
import kotlinx.serialization.Serializable
import kotlin.time.Duration

@Serializable
data class VerifierMetrics(
    val compilationDuration: Duration = Duration.ZERO,
    val backAnnotationDuration: Duration = Duration.ZERO,
    val portfolioDuration: Duration = Duration.ZERO,
)

@Serializable
data class VerificationMetrics(
    val totalDuration: Duration = Duration.ZERO,
    val backend: BackendMetrics? = null,
    val verifier: VerifierMetrics = VerifierMetrics(),
)
